Document the new dependency resolution algorithm (#3340)
NuGet/docs.microsoft.com-nuget: docs/consume-packages/Package-References-in-Project-Files.md

| `-LockFilePath` | `--lock-file-path` | NuGetLockFilePath | Defines a custom lock file location for a project. By default, NuGet supports `packages.lock.json` at the root directory. If you have multiple projects in the same directory, NuGet supports project specific lock file `packages.<project_name>.lock.json` |

## NuGet Dependency Resolver

The NuGet dependency resolver follows the [4 rules as described in the dependency resolution document](../../docs/concepts/Dependency-Resolution.md).

In order to improve the performance and scalability of the restore operation, the restore algorithm was rewritten in the 6.12 release.
As of the 6.12 release, the new restore algorithm is enabled by default for all PackageReference projects.
While the new restore algorithm is is functionally equivalent to the previous one, as with any software, bugs are possible.
To revert to the previous implementation, set the MSBuild property `RestoreUseLegacyDependencyResolver` to `true`.

Should you face restore failures in 6.12, .NET 9 or 17.12, that weren't reproducing in earlier versions, please [file an issue on GitHub](https://github.com/NuGet/Home/issues/).

> Any potential discrepancies may manifest in different ways, such as during compilation or at runtime. There's also a chance that changes don't lead to failures, but just differences.

Reviewer: ```suggestion
Any potential discrepancies may manifest in different ways, such as during compilation or at runtime.
There's also a chance that changes don't lead to failures, but just differences.
```

This paragraph was almost, but not quite, one sentence per line 😁

> Any potential discrepancies may manifest in different ways, such as during compilation or at runtime. There's also a chance that changes don't lead to failures, but just differences.

Reviewer: from https://learn.microsoft.com/en-us/contribute/content/style-quick-start#use-the-microsoft-voice-principles:

> Use everyday words: Try to use natural language, the words your customers use. Be less formal but not less technical. Provide examples that explain new concepts.

I think something like this is better:

```suggestion
Any differences between the old and new algorithms may have different impacts, such as during compilation or at runtime. There's also a chance that changes don't lead to failures, but different package versions being restored.
```

If you think you may be impacted by any changes, here are the steps you can take to verify whether the changes in the NuGet restore algorithm are the root cause.

> Restore's intermediates generated in the `MSBuildProjectExtensionsPath` folder can help determine the equivalence between the 2 algorithms. Usually this is the `obj` folder of your build. You can use `msbuild.exe` or `dotnet.exe` for the next steps.

Reviewer: ```suggestion
Restore's intermediates generated in the `MSBuildProjectExtensionsPath` folder can help determine the equivalence between the 2 algorithms.
Usually this is the `obj` folder of your build.
You can use `msbuild.exe` or `dotnet.exe` for the next steps.
```

> Restore's intermediates generated in the `MSBuildProjectExtensionsPath` folder can help determine the equivalence between the 2 algorithms. Usually this is the `obj` folder of your build. You can use `msbuild.exe` or `dotnet.exe` for the next steps.

Reviewer: again, as per the writing style's guidelines of being more casual, every day language:
```suggestion
Restore write its results in the `MSBuildProjectExtensionsPath` directory, which can be compared with the new and old algorithms to find differences. Usually this is the `obj` folder of your build. You can use `msbuild.exe` or `dotnet.exe` for the next steps.
```

1. Remove the `obj` folder for your project.
1. Run `msbuild -t:restore`
1. Save the contents of the `obj` to a location indicating that it's the `new` behavior.
1. Run `msbuild -t:restore -p:RestoreUseLegacyDependencyResolver="true"`
1. Save the contents of the `obj` to a location indicating that it's the `legacy` behavior.

> 1. Analyze the files yourself, or create a NuGet/Home issue and we'll take a look.

Reviewer: ```suggestion
1. Compare the files in the two directories, particularly *project.assets.json*.
   Tools that can highlight differences are especially useful for this (for example, Visual Studio Code, open both files, and use the right-click "select for compare" and "compare to selected")
```
